packages/toolkit/src/tsHelpers.ts

```diff
@@ -207,3 +207,11 @@ export type Tail<T extends any[]> = T extends [any, ...infer Tail]
   : never
 
 export type UnknownIfNonSpecific<T> = {} extends T ? unknown : T
+
+/**
+ * A Promise that will never reject.
+ * @see https://github.com/reduxjs/redux-toolkit/issues/4101
+ */
+export type SafePromise<T> = Promise<T> & {
+  __brand: 'SafePromise'
```

Comment:
Is this already set in stone? I fear that `__brand` could get collisions at some point in time - what about `__eslint: 'SafePromise'`?

Reply:
Oh and - in case it's not clear from the issue, my current proposal is that each library define its own brand. Not that we have a shared built-in in the ts-eslint plugin for folks to use. This means users would have to manually configure the `plugin:@typescript-eslint/no-floating-promises` allowlist for each library with "safe" Promises:

```js
"@typescript-eslint/no-floating-promises": ["error", {
  "allowKnownSafePromises": [
    { "from": "package", "name": "SafePromise", "package": "@reduxjs/toolkit"
  ]
}]
```

Reply:
Oh and - in case it's not clear from the issue, my current proposal is that each library define its own brand. Not that we have a shared built-in in the ts-eslint plugin for folks to use. This means users would have to manually configure the `plugin:@typescript-eslint/no-floating-promises` allowlist for each library with "safe" Promises:

```js
"@typescript-eslint/no-floating-promises": ["error", {
  "allowKnownSafePromises": [
    { "from": "package", "name": "SafePromise", "package": "@reduxjs/toolkit"
  ]
}]
```

Another strategy would be asking end libraries to take a dependency on some package -likely a `@typescript-eslint/*`- one in order to mark their "safe" Promises. I don't think that's an ask we can put forward. And even if we did, folks will want a way to add an allowlist anyway.

Reply:
> And I'd expect it wouldn't be feasible for most user-facing libs to take a types dependency on any @typescript-eslint/* package.

That's why I added the `@ts-ignore` - if it's installed and available, use it, if not, not.

> Oh and - in case it's not clear from the issue, my current proposal is that each library define its own brand. 

Ah, yeah, that wasn't clear. Yeah, I guess we can explore it that way and we'll see if some pattern will stick out in the end :)

`__safetyBrand` seems fine so far. I'll try one alternative suggestion, though: `__linterBrands`. 
I think this could be defined as a string union, so if any other linter-related brands would come up in the future, they could just add more like `__linterBrands: 'SafePromise' | 'StableValue'`, right?

Reply:
`__linterBrands` it is! 🚀